### DISCUSSION TITLE
setLoop for MP3 Encoded files

### DISCUSSION THREAD

**Discussion by RDJunc** [pschatzmann/arduino-audio-tools#1094] [Q&A] [answered]
I have looked through as much of the posts and documentation as I could but without success.  I have a larger program that operates just fine but I want to implement the playing of a bell sound.  It will be initiated with an event and should continue to play until other logic ends it.

For testing, I have used your example for streaming an MP3 file to i2s.  It works wonderfully but plays the file just once.  The code included here was my attempt to get it to play continuously.  No joy.

It is my understanding that there is no setLoop() for this class.  If so, what is the best way to get the file to play continuously?  Will it be necessary for me to convert the file to raw and use the MemoryStream method?  If so, I have to figure out a way to convert the file to the raw format.

Thanks in advance for the assistance.

```
/**
 * @file streams-memory_mp3_short-i2s.ino
 * @author Phil Schatzmann
 * @brief decode MP3 stream and output as i2s signal
 * @version 0.1
 * @date 2021-01-24
 *
 * @copyright Copyright (c) 2021
 */

#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "RRCrossingBell02.h"

MemoryStream mp3(RRCrossingBell02_mp3, RRCrossingBell02_mp3_len);
I2SStream i2s;
MP3DecoderHelix helix;
EncodedAudioStream out(&i2s, &helix); // output to decoder
StreamCopy copier(out, mp3);    // copy in to i2s

void setup() {
    Serial.begin(115200);
    //AudioLogger::instance().begin(Serial, AudioLogger::Info);  

    // begin processing
    auto cfg = i2s.defaultConfig();
    cfg.sample_rate = 24000;
    cfg.channels = 1;
    cfg.pin_data = 25;
    cfg.pin_bck = 32;
    cfg.pin_ws = 33;

    i2s.begin(cfg);
    out.begin();
    
}

void loop() {

    copier.copy();

    if (!copier.copy()) {
        i2s.begin();
        out.begin();
    }

}
```

**Comment by pschatzmann** [comment, marked as answer]
Have a look at the FileLoop class.
Alternatively you can just move the file pointer to the beginning when you are at the end.

**Reply by RDJunc** [reply]
Thanks for the assistance.  I couldn't figure out how to use the FilLoop class since the I am playing the file from memory.  (I could not get the syntax correct for FileLoop.setFile().)

So I tried your suggestion for reseting the mp3 file pointer.  Below is how I did it.  It may not be the best way though.  This is a bell sound that repeats.  The bell rings five times in quick succession.  This keeps the file very small.  The downside is that with the method I used to restart the playing of the file, there is noticeable pause each time the file replays.  Is there a way make that restart faster?   Is the FileLoop class faster and if so, how do I either load it from memory or from the same folder as the project.

Thanks again very much for the help and these very powerful tools.  As I am fond of saying, "Just because someone has a book and a knife, that doesn't make them a surgeon."  I freely admit I am no surgeon.

```
/**
 * @file streams-memory_mp3_short-i2s.ino
 * @author Phil Schatzmann
 * @brief decode MP3 stream and output as i2s signal
 * @version 0.1
 * @date 2021-01-24
 *
 * @copyright Copyright (c) 2021
 */

#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "RRCrossingBell02.h"

MemoryStream mp3(RRCrossingBell02_mp3, RRCrossingBell02_mp3_len);
I2SStream i2s;
MP3DecoderHelix helix;
EncodedAudioStream out(&i2s, &helix); // output to decoder
StreamCopy copier(out, mp3);    // copy in to i2s

void setup() {
    Serial.begin(115200);
    //AudioLogger::instance().begin(Serial, AudioLogger::Info);  

    // begin processing
    auto cfg = i2s.defaultConfig();
    cfg.sample_rate = 24000;
    cfg.channels = 1;
    cfg.pin_data = 25;
    cfg.pin_bck = 32;
    cfg.pin_ws = 33;

    i2s.begin(cfg);
    out.begin();

}

void loop() {

    if (mp3.available()) {
        copier.copy();
    }
    else {
        mp3.begin();
        
        //helix.end(); // flush output
        //auto info = out.decoder().audioInfo();
        //LOGI("The audio rate from the mp3 file is %d", info.sample_rate);
        //LOGI("The channels from the mp3 file is %d", info.channels);
        //i2s.end();
        //stop();
    }
}
```

**Reply by codegrue** [reply]
Wanted to add if the mp3 is a SPIFFS file, this works:

```
File mp3;

if (mp3.available())
    copier.copy();
else
    mp3.seek(0);
```

**Comment by pschatzmann** [comment]
In case you did not find it: here is the [link to the example](https://github.com/pschatzmann/arduino-audio-tools/blob/bdc3713fd30878b196e381d1b462c35ecd77b768/examples/examples-audiokit/streams-file_loop-audiokit/streams-file_loop-audiokit.ino#L15)

MP3 needs a little bit of time to get started, so if you want to avoid any gaps you could use a wav file instead...

**Reply by RDJunc** [reply]
I did see that example by my application will not have a SD card.  The sound clip will be in the ESP32 memory.  Is there an equivalent to the FileLoop class sounds being played from memory?  I will try using a wav file.

**Comment by pschatzmann** [comment]
I have access to a PC now and I had a look at your first example. I think your description was confusing me because you were talking about about a file.

Look at the documentation for the [MemoryStream](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_memory_stream.html)! Why don't you just call mp3.setLoop(true) ?

The FileLoop is to be used for any type of file including SPIFFS or LittleFS which do not require any SD drive

**Reply by RDJunc** [reply]
I converted the file to a wav and used setLoop and it worked.  I will go back to using the mp3 version since it is so much smaller.  Now the issue is a popping sound that happens as the sound repeats.   This is a known issue with the MAX9835 and can be dealt with through the use of the SD_MODE line to the device.  Unfortunately, the testing breakout board I am using does not bring the SD_MODE line out but my design does.  I'll need to check to see how or if your library supports that.  Thanks again for your help.

**Reply by pschatzmann** [reply]
You could try if a MedianFilter is of any help

**Reply by RDJunc** [reply]
Interesting.  When I used the mp3 version of the sound, there was the delay that you mentioned but the pop very slight.  When I went back to the wav, there was no delay but the pop was more pronounced.

The sound source I am using for the bell never goes off completely.  The next bell sound begins before the previous one has completely decayed.  This means that when the file repeats, the level at the end of the file needs to closely match the level at the beginning.  I have done that to some degree.  I may have to spend some time with the file and get the each bell ring to decay completely so that when the file loops, there is a perfect match.

The bell sound is the sound that you hear at a railroad crossing.

**Reply by RDJunc** [reply]
You have been very helpful and I have (I believe) just one more request.

I am running into a bit of a memory management issue.  In most cases my sketches have had no problem fitting into the default set of memory with everything loading into DRAM.  With the one sound clip I am using and keeping the file size and sampling rate low, everything fits but there is plenty of additional memory available.

For the first time, I am using SPIFFs.  The default of 1.5M is fine.  I can actually load several sound clips into the space.  I have successfully added the .h file for the wave clip but am struggling to figure out the right way to bring it in from the SPIFFs so that it will play through your tools.  (This might be a good example to add to your already comprehensive list.)

Thanks again.

**Reply by RDJunc** [reply]
Whoops.  I may have found your example under the Player examples.

**Reply by pschatzmann** [reply]
Or you can take also any other file examples which are not based on the player. After all, a file is just a subclass of an arduino stream, so you can take them as copy source...
